From c9d75cc5dc40e127701de2ff32fcc2fa6c49a7a5 Mon Sep 17 00:00:00 2001
From: Gajanan Bhat <gajanan.bhat@intel.com>
Date: Wed, 2 Dec 2015 14:36:26 +0530
Subject: [PATCH 1/6] SQUASHME [VPG]: drm/i915: Check if Panel fitter width and
 height are set to zero

Setting src_w and src_h values of panel fitter to zero is creating divide by zero
error. Add check before using these values.

SQUASHME: This patch should be squashed with below patch:
commit 736a850003c9364569cad37a9e98afebefcfc207
Author: Arun R Murthy <arun.r.murthy@intel.com>
Date:   Tue Aug 11 14:29:54 2015 +0530

    SQUASHME! [VPG]: calculate the aspect ratio for pfit

Change-Id: Ia489565e8f4fc4a186c2e2c72eea64b3d282e6a0
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-10776
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
Reviewed-on: https://android.intel.com:443/463650
---
 drivers/gpu/drm/i915/intel_display.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 829028a..b303643 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11239,6 +11239,15 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 			DRM_ERROR("Not enabling Panel Fitter\n");
 		} else {
 			u32 pfit_mode = 0;
+			if (disp->panel_fitter.mode == PFIT_OFF ||
+					disp->panel_fitter.src_w == 0 ||
+					disp->panel_fitter.src_h == 0) {
+				pfit_control &= ~PFIT_ENABLE;
+				intel_crtc->scaling_src_size =
+						(((mode->hdisplay - 1) << 16) |
+							(mode->vdisplay - 1));
+				goto pfit_out;
+			}
 
 			pfit_control &= ~PFIT_PIPE_MASK;
 			pfit_control |= (intel_crtc->pipe << PFIT_PIPE_SHIFT);
@@ -11246,8 +11255,7 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 			intel_crtc->scaling_src_size =
 				(((disp->panel_fitter.src_w - 1) << 16) |
 						(disp->panel_fitter.src_h - 1));
-			if (disp->panel_fitter.mode == PFIT_OFF);
-			else if (((mode->hdisplay * disp->panel_fitter.src_h) /
+			if (((mode->hdisplay * disp->panel_fitter.src_h) /
 						disp->panel_fitter.src_w) < mode->vdisplay)
 				pfit_mode |= PFIT_SCALING_LETTER;
 			else if (((mode->vdisplay * disp->panel_fitter.src_w) /
@@ -11293,6 +11301,7 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 				pfit_control &=  MASK_PFIT_SCALING_MODE;
 				pfit_control |= pfit_mode;
 			}
+pfit_out:
 			intel_crtc->pfit_control = pfit_control;
 			if (pfit_control != pfitcontrol || scaling_src_size != intel_crtc->scaling_src_size)
 				dev_priv->pfit_changed = true;
-- 
1.9.1

